### DISCUSSION TITLE
"traverse:holds(obj)" lock is not letting me through, even with the object

### DISCUSSION THREAD

**Discussion by RodRitter** [evennia/evennia#2661] [Evennia Questions] [answered]
I'm trying to create a key that allows anyone who holds it, into a specific room. I'm not sure if I got the lock wrong, but any help is appreciated.

```
# shop.py

# Create storeroom
self.db.storeroom = create_object(typeclass=Room, key="Storeroom")

# Create a key
storeroom_key = create_object(typeclass=RoomKey, key=key_name, home=self)

# Create an exit that goes to "storeroom"
storeroom_entry = create_object(typeclass=Exit, key="storeroom", home=self, destination=new_storeroom)
storeroom_entry.location = self
storeroom_entry.locks.add(f"traverse:holds({storeroom_key})") # I've set a lock here
```

`examine storeroom ` - the Exit, returns this:
```
Locks: 
  call:true(); control:perm(Developer); delete:perm(Admin); drop:holds();
  edit:perm(Admin); examine:perm(Builder); get:false(); puppet:false();
  tell:perm(Admin); traverse:holds(Storeroom Key (747)); view:all()
```
I'm holding the key at this point. `examine key` - returns this:
```
Name/key: Storeroom Key (747) (#748)
Typeclass: RoomKey (typeclasses.objects.RoomKey)
Location: Barker (#1)
```
When I try to access storeroom, I cannot access it:
```
> quell
> storeroom
> You cannot go there.
```

Am I doing something wrong?

**Comment by InspectorCaracal** [comment, marked as answer]
> ```
> Locks: 
>   call:true(); control:perm(Developer); delete:perm(Admin); drop:holds();
>   edit:perm(Admin); examine:perm(Builder); get:false(); puppet:false();
>   tell:perm(Admin); **traverse:holds(Storeroom Key (747));** view:all()
> ```

I might be wrong but shouldn't that be `traverse:holds(#748)` in order to work? I think you need to explicitly use the dbref instead of just the object variable when creating your lockstring

**Reply by Griatch** [reply]
@RodRitter You can't just dump the object in there - the lock is a string! Either store `storeroom_key.key` or `storeroom_key.dbref` in the lock.

@InspectorCaracal The `holds()` lockfunc accepts the object's `key` as well as its dbref.

**Reply by RodRitter** [reply]
Yep! that did it. I feel stupid for not trying that. Thanks for the reply, lesson learned :)

**Reply by RodRitter** [reply]
@Griatch I initially used the key but it didn't work. Like the npc shop tutorial.

**Reply by Griatch** [reply]
@RodRitter Just as an example, I tested the following successfully in-game:

    > create/drop key
    > create/drop box 
    > lock box = get:holds(key)         # specified by key rather than #dbref
    > quell                             # (since I was superuser)
    > get box 
    You can't get that 
    > get key 
    You pick up key.
    > get box 
    You pick up box. 

As seen, once I had the `key` in hand, I could also pick up the box.

**Reply by RodRitter** [reply]
@Griatch That's strange, then why would it not work for me if the lock specifically had `traverse:holds(Storeroom Key (747))` instead of db_ref? 

Is it different in a class, than it is in a command?

**Reply by aMiss-aWry** [reply]
Storeroom Key (747) is meaningless to the parser. Er... Unless the literal object name is 'Storeroom Key (747)' I guess.

**Reply by Griatch** [reply]
Because what you store there is a string representation of the object (equivalent to doing `repr(storeroom_key)`) and _not_ the name of the object (which is stored in the `storeroom_key.key` field).

**Reply by RodRitter** [reply]
Thanks for the explanations everyone. I understand now. In my example, what was shown in `examine` is not the key, but rather the object itself.
